### ISSUE TITLE
Expose port in Dockerfiles. Gitlab CI and Github Actions do health checks based on those.

### ISSUE THREAD

**Issue by lsjostro** [GoogleCloudPlatform/cloud-sql-proxy#1544]
## Bug Description
When running a cloud sql proxy container as a "service container" in Gitlab CI or Github Actions they do health check on the `EXPOSE` port in the container. So when running the upstream container as a service it will fail to start due to missing EXPOSE keyword in Dockerfile. We should have a default expose value in the Dockerfiles. 

What about adding this to the Dockerfiles, which will solve running the official container images as "service containers" in gitlab/github CI? 

```
EXPOSE 5432 3306 1433
```

**Comment by enocom** [comment]
Could the person who configures the service container map these ports themselves?

For example:

https://docs.github.com/en/actions/using-containerized-services/about-service-containers#example-mapping-redis-ports

**Comment by kurtisvg** [comment]
FWIW, I think exposing a port is a bad idea since the port is configurable by the user -  I don't think we should expose ports not in use.

I'm also not sure that GitlabCI or Github Actions would correctly healthcheck a TCP port to a SQL instance.

**Comment by lsjostro** [comment]
> Could the person who configures the service container map these ports themselves?
> 
> For example:
> 
> https://docs.github.com/en/actions/using-containerized-services/about-service-containers#example-mapping-redis-ports
In Gitlab CI it's not possible I'm afraid. 😞  I'm using a custom dockerfile like so (which is a bit annoying to publish and maintain):
```
FROM <offical sql proxy container image>
EXPOSE 5432
```

**Comment by lsjostro** [comment]
> FWIW, I think exposing a port is a bad idea since the port is configurable by the user - I don't think we should expose ports not in use.
> 
> I'm also not sure that GitlabCI or Github Actions would correctly healthcheck a TCP port to a SQL instance.

EXPOSE is mainly used for inter-communication inside docker network, not really widely used anymore. So it doesn't really cause any issues having those entries in the dockerfile? Can be good as documentation to which ports the container can potentially expose?

me googling about drawbacks using `EXPOSE` and found this source. https://we-are.bookmyshow.com/understanding-expose-in-dockerfile-266938b6a33d

**Comment by enocom** [comment]
The difficult thing here is that customers use all sorts of different ports and so we don't have a good default. There are built-in health-checks on a HTTP port, but those have to be enabled.

Seems like this is a problem with Gitlab especially. GitHub at least lets a person configure the exposed ports.

**Comment by lsjostro** [comment]
yeah, but they will be overridden with the -p flag. I don't seem any harm adding this. Adding the default ports for Postgres, mysql and mssql I think make sense (engines google supports). I mean `redis`, `postgres`, `mysql` etc. all have `EXPOSE` in their offical container images, as well as the offical `nginx` and `envoy`-proxy which has `EXPOSE` and customer of those will for sure use different ports.

**Comment by enocom** [comment]
Thinking some more about this, I see that Gitlab will wait for all exposed ports to be accessible. So if someone didn't have an instance listening on any of the default engine ports, would the current PR work?

https://docs.gitlab.com/ee/ci/services/#how-the-health-check-of-services-works

**Comment by lsjostro** [comment]
@enocom yeah, you're right! Multiple ports won't work! 😢 Either we blame Gitlab to fix their stuff or we give our customers at least one exposed port which can be manually used? Less confusing would probably be not adding an exposed port I guess? What do you guys think? 🤷🏻

**Comment by lsjostro** [comment]
I'm just leaving this here for reference if someone runs into the same issue. Seems that gitlab at least has it one it's radar. 
https://gitlab.com/gitlab-org/gitlab-runner/-/issues/2460

**Comment by enocom** [comment]
This does seem like a problem with Gitlab. I'd love to do something to help here, but without having a good port to expose for everyone, there's not much that we can do.

In this case maintaining a custom container that exposes that ports a person actually uses is probably the best solution, in spite of the overhead.

**Comment by lsjostro** [comment]
Totally! let's close this one! Hope it can help someone else maybe.

**Comment by enocom** [comment]
Feel free to re-open or update if anything changes on Gitlab's side.
